### PR TITLE
Fix TRTLLM ragged prefill edge cases

### PR DESCRIPTION
**Pull request by alexeldeib**
## 📌 Description

I found my way here while debugging https://github.com/vllm-project/vllm/issues/42426 which eventually brought me to #3047 which seemed very similar to my issue, and porting a similar fix to the trtllm ragged path solved it for me. This approach uses a wrapper since the underlying kernel appears to still be closed source despite the variety of OSS kernel dumps to flashinfer.

I had a similar branch to vllm https://github.com/vllm-project/vllm/compare/main...alexeldeib:ace/trtllm-ragged-empty-row-compact-no-lse but since #3047 was originally affecting sglang I thought it might make sense to take it here instead

also i'm a kernel noob so forgive any basic errors, this was AI assisted, but conceptually it seems to align with prior art/bugs, and it does reliably resolve my issue in vllm (happy to share more details but I figure 3047 is already validation of the theory in some part)

---

TRTLLM-GEN ragged attention packs non-paged SeparateQkv K/V rows along a single token axis. Do not pass a synthetic batch stride into the TMA descriptor for that singleton-batch layout; use zero stride instead so both negative and positive int32 wraparound cases are avoided.

The Python wrapper now defines the empty-attention identity for rows that have query tokens but no current KV tokens: out=0 and lse=-inf. It launches the kernel only for active rows and scatters active results back into the original output buffers, avoiding undefined empty-KV kernel behavior.

Optional trusted CPU length mirrors let framework callers keep the all-active fast path asynchronous. Direct callers without mirrors still derive lengths from device indptrs and take the correctness-preserving compaction path when empty rows are present. CUDA regression coverage exercises stride overflow, positive stride wraparound, mixed empty-KV rows, all-empty rows, direct-caller fallback, and invalid CPU length mirrors.

Note, that last part does change the signature and may be an issue, but might be useful in lieu of modifying the underlying kernel to handle this directly?

## 🔍 Related Issues

Related: https://github.com/flashinfer-ai/flashinfer/issues/3047. This is the TRTLLM-GEN empty-KV case called out in https://github.com/flashinfer-ai/flashinfer/issues/3047#issuecomment-4473332762, distinct from the FA2/FA3 path fixed in https://github.com/flashinfer-ai/flashinfer/pull/3251.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Added optional CPU-side Q/KV sequence-length mirrors to compact ragged batches and efficiently handle empty KV rows.
* **Bug Fixes**
  * Corrected K/V batch-stride handling for SeparateQkv layouts.
  * Ensured empty KV rows produce zero outputs and `-inf` log-sum-exp values.
  * Added validation requiring CPU length mirrors during CUDA graph capture.
* **Tests**
  * Expanded coverage for stride behavior, empty-row semantics, CPU-mirror equivalence, and CUDA graph capture.

### CODE REVIEW

File: flashinfer/prefill.py [hunk 1]
```diff
@@ -4040,45 +4050,224 @@ def trtllm_ragged_attention_deepseek(
...
+                        torch.zeros(
+                            1, device=query.device, dtype=cum_seq_lens_kv.dtype
+                        ),
+                        torch.cumsum(run_seq_lens, dim=0, dtype=cum_seq_lens_kv.dtype),
+                    ],
+                    dim=0,
+                )
+                run_max_q_len = int(run_q_lens.max().item())
+                run_max_kv_len = int(run_seq_lens.max().item())
+                run_batch_size = int(active_indices.numel())
```
[the remark this answers is not in the document]
Comment (alexeldeib) [reply flashinfer/prefill.py]: seems reasonable, will adjust + revalidate

[the remark this answers is not in the document]
Comment (alexeldeib) [reply flashinfer/prefill.py]: Addressed in the current branch state. The CPU mirror path now performs the compaction bookkeeping on CPU: active-row detection, active/inactive index selection, compacted lengths, cumulative lengths, and max lengths are all derived from the CPU mirrors before copying the compacted metadata to the device.

File: flashinfer/prefill.py [hunk 2]
```diff
@@ -4220,6 +4222,14 @@ def trtllm_ragged_attention_deepseek(
...
     backend : str
         Attention backend to use. "trtllm-gen" (default) or "cute-dsl".
+    q_seq_lens_cpu : Optional[torch.Tensor]
+        Optional trusted CPU mirror of the per-row query lengths. When provided
+        together with ``kv_seq_lens_cpu``, the Python wrapper can keep the
+        all-active ragged fast path asynchronous while still compacting empty-KV
+        rows. If omitted, the wrapper derives lengths from the device indptrs
+        and may synchronize to preserve correctness for direct callers.
+    kv_seq_lens_cpu : Optional[torch.Tensor]
+        Optional trusted CPU mirror of the per-row KV lengths.
```
[the remark this answers is not in the document]
Comment (alexeldeib) [reply flashinfer/prefill.py]: Addressed in 6dbd301f. The q_seq_lens_cpu and kv_seq_lens_cpu docstrings now explicitly scope these CPU mirrors to the trtllm-gen backend.

File: flashinfer/prefill.py [hunk 3]
```diff
@@ -4335,45 +4345,312 @@ def trtllm_ragged_attention_deepseek(
...
+            if not bool(active_rows_cpu.all().item()):
+                has_inactive_rows = True
+                has_active_rows = bool(active_rows_cpu.any().item())
+        else:
+            q_lens = cum_seq_lens_q[1:] - cum_seq_lens_q[:-1]
+            kv_lens = cum_seq_lens_kv[1:] - cum_seq_lens_kv[:-1]
+            active_rows = (q_lens > 0) & (kv_lens > 0)
+            has_inactive_rows = bool((~active_rows).any().item())
+            if has_inactive_rows:
+                has_active_rows = bool(active_rows.any().item())
```
[the remark this answers is not in the document]
Comment (alexeldeib) [reply flashinfer/prefill.py:4751]: Addressed in 6dbd301f. The no-CPU-mirror fallback now raises a clear ValueError during CUDA graph capture, and the test coverage includes that guard plus the all-empty direct fallback case.

File: csrc/trtllm_fmha_kernel_launcher.cu [hunk 4]
```diff
@@ -740,10 +740,10 @@ void trtllm_ragged_attention(
   int head_dim_v = value.size(2);
   int k_stride_keys_values = key.stride(0);
   int k_stride_heads = key.stride(1);
-  int k_stride_batch = key.numel();
+  int k_stride_batch = 0;
```
Comment (PerkzZheng) [inline csrc/trtllm_fmha_kernel_launcher.cu:818]: can you also help add a comment as kernelParams.h even though they seems to be duplicated. others LGTM.

**Review by PerkzZheng — APPROVED** [review]

File: csrc/trtllm_fmha_kernel_launcher.cu (hunk 4 above)
Comment (alexeldeib) [reply csrc/trtllm_fmha_kernel_launcher.cu:818]: Added the launcher-side comment in the latest push. It now mirrors the kernelParams.h rationale that ragged SeparateQkv K/V is packed along one token axis, so the generated TMA shape uses a singleton batch dimension and the launcher should pass zero batch stride instead of a synthetic numel-derived stride.

**Review by PerkzZheng — APPROVED** [review]

**Comment by alexeldeib** [comment]
anything I can do here to get CI to run/get more approvals? happy to rebase to keep it up to date but the repo moves quickly and CI can't run since i'm not an org member?

not sure what to do

**Comment by aleozlx** [comment]
adding @saltyminty and @qsang-nv to review and follow up on this

**Review by saltyminty — APPROVED** [review]
Approved pending CI
